src/pretix/base/exporters/invoices.py

```diff
@@ -84,43 +133,243 @@ def render(self, form_data: dict, output_file=None):
+                    _('Yes') if i.reverse_charge else _('No'),
+                    i.foreign_currency_display,
+                    i.foreign_currency_rate,
+                    i.total_gross if i.total_gross else Decimal('0.00'),
+                    Decimal(i.total_net if i.total_net else '0.00').quantize(Decimal('0.01')),
+                    pmi,
+                    ', '.join([
+                        str(self.providers.get(p, p)) for p in sorted(set((i.payment_providers or '').split(',')))
+                        if p and p != 'free'
+                    ])
+                ]
+        elif sheet == 'lines':
+            yield [
+                _('Invoice number'),
+                _('Line number'),
+                _('Description'),
+                _('Gross price'),
+                _('Net price'),
+                _('Tax value'),
+                _('Tax rate'),
+                _('Tax name'),
+                _('Event start date'),
+
+                _('Date'),
+                _('Order code'),
+                _('E-mail address'),
+                _('Invoice type'),
+                _('Cancellation of'),
+                _('Invoice sender:') + ' ' + _('Name'),
+                _('Invoice sender:') + ' ' + _('Address'),
+                _('Invoice sender:') + ' ' + _('ZIP code'),
+                _('Invoice sender:') + ' ' + _('City'),
+                _('Invoice sender:') + ' ' + _('Country'),
+                _('Invoice sender:') + ' ' + _('Tax ID'),
+                _('Invoice sender:') + ' ' + _('VAT ID'),
+                _('Invoice recipient:') + ' ' + _('Company'),
+                _('Invoice recipient:') + ' ' + _('Name'),
+                _('Invoice recipient:') + ' ' + _('Street address'),
+                _('Invoice recipient:') + ' ' + _('ZIP code'),
+                _('Invoice recipient:') + ' ' + _('City'),
+                _('Invoice recipient:') + ' ' + _('Country'),
+                _('Invoice recipient:') + ' ' + pgettext('address', 'State'),
+                _('Invoice recipient:') + ' ' + _('VAT ID'),
+                _('Invoice recipient:') + ' ' + _('Beneficiary'),
+                _('Invoice recipient:') + ' ' + _('Internal reference'),
+                _('Payment providers'),
+            ]
+
+            p_providers = OrderPayment.objects.filter(
+                order=OuterRef('invoice__order'),
+                state__in=(OrderPayment.PAYMENT_STATE_CONFIRMED, OrderPayment.PAYMENT_STATE_REFUNDED,
+                           OrderPayment.PAYMENT_STATE_PENDING, OrderPayment.PAYMENT_STATE_CREATED),
+            ).values('order').annotate(
+                m=GroupConcat('provider', delimiter=',')
+            ).values(
+                'm'
+            ).order_by()
+
+            qs = InvoiceLine.objects.annotate(
+                payment_providers=Subquery(p_providers, output_field=CharField()),
+            ).filter(
+                invoice__in=self.invoices_queryset(form_data)
```

Comment:
I was a little worried that using a subquery here might have worse performance characteristics than the join we previously used, considering we might realistically have millions of invoices in a cross-event export of a large client.

However, apparently at least PostgreSQL will be smart about it:
https://www.percona.com/blog/2020/04/16/sql-optimizations-in-postgresql-in-vs-exists-vs-any-all-vs-join/

I'm still a little worried about this on other databases, but since it makes the code much simpler I think we can just give it a go and see if it causes actual problems.

Reply:
I also thought about caching the invoices from when the invoices sheet is rendered. That way we wouldn't need to have that query twice. But that would need state in the exporter/a change in how the Multi Sheet Exporter works.